### PR TITLE
`x509_info_subject_alt_name`: Render HardwareModuleName as hex

### LINKED ISSUE #6262: `mbedtls_x509_crt_info` produces non-printable output for binary HardwareModuleName SAN extensions
### Summary

`mbedtls_x509_crt_info` produces non-printable characters if the certificate contains an RFC 4108 HardwareModuleName (as a Subject Alternative Name extension) whose hwSerialNum has non-printable bytes.

This happens because [the code](https://github.com/Mbed-TLS/mbedtls/blob/52f83dc471625f503a335ccef4f4b484a2a579ef/library/x509_crt.c#L1915) assumes the hwSerialNum is made of printable characters and simply does a `memcpy` directly to the output. In fact, RFC 4108 specifies that the ASN.1 type of hwSerialNum is OCTET STRING, which basically permits any arbitrary sequence of octets. The RFC does not impose any other structure or encoding on the serial number.

### System information

Mbed TLS version (number or commit id): 52f83dc471625f503a335ccef4f4b484a2a579ef
Operating system and version: Manjaro Linux 21.3.7
Configuration (if not default, please attach `mbedtls_config.h`): default
Compiler and options (if you used a pre-built binary, please indicate how you obtained it): gcc (GCC) 12.1.1 20220730, default options
Additional environment information: None

### Expected behavior

In order to print binary hwSerialNums, we'd have to encode them into a printable format (hex for instance).

I generated this test certificate with a binary hwSerialNum (3132338081008180333231 in hex):
```
-----BEGIN CERTIFICATE-----
MIIBwTCCAWagAwIBAgIBTTAKBggqhkjOPQQDAjBPMQswCQYDVQQGEwJVSzERMA8G
A1UECgwITWJlZCBUTFMxLTArBgNVBAMMJE1iZWQgVExTIG5vbi1wcmludGFibGUg
b3RoZXJuYW1lIFNBTjAeFw0yMjA5MDYxNTU2NDdaFw0zMjA5MDMxNTU2NDdaME8x
CzAJBgNVBAYTAlVLMREwDwYDVQQKDAhNYmVkIFRMUzEtMCsGA1UEAwwkTWJlZCBU
TFMgbm9uLXByaW50YWJsZSBvdGhlcm5hbWUgU0FOMFkwEwYHKoZIzj0CAQYIKoZI
zj0DAQcDQgAEN8xW2XYJHlpyPsdZLf8gbu58+QaRdNCtFLX3aCJZYpJO5QDYIxH/
6i/SNF1dFr2KiMJrdw1VzYoqDvoByLTt/6MzMDEwLwYDVR0RBCgwJqAkBggrBgEF
BQcIBKAYMBYGBysGAQQBEQMECzEyM4CBAIGAMzIxMAoGCCqGSM49BAMCA0kAMEYC
IQDATir07PTj5gtf+HAyI+nd27AH9+bdaWdOI2t2bAwUWgIhAO7kvdcsa++yfJdT
3vnWdvcHRIAdXA0kh+mcBMaXk9B0
-----END CERTIFICATE-----
```

The expected output of `mbedtls_x509_crt_info` should then be:
```
cert. version     : 3
serial number     : 4D
issuer name       : C=UK, O=Mbed TLS, CN=Mbed TLS non-printable othername SAN
subject name      : C=UK, O=Mbed TLS, CN=Mbed TLS non-printable othername SAN
issued  on        : 2022-09-06 15:56:47
expires on        : 2032-09-03 15:56:47
signed using      : ECDSA with SHA256
EC key size       : 256 bits
subject alt name  :
    otherName :
        hardware module name :
            hardware type          : 1.3.6.1.4.1.17.3
            hardware serial number : 3132338081008180333231
```

### Actual behavior

The actual output of `mbedtls_x509_crt_info` is
```
cert. version     : 3
serial number     : 4D
issuer name       : C=UK, O=Mbed TLS, CN=Mbed TLS non-printable othername SAN
subject name      : C=UK, O=Mbed TLS, CN=Mbed TLS non-printable othername SAN
issued  on        : 2022-09-06 15:56:47
expires on        : 2032-09-03 15:56:47
signed using      : ECDSA with SHA256
EC key size       : 256 bits
subject alt name  :
    otherName :
        hardware module name :
            hardware type          : 1.3.6.1.4.1.17.3
            hardware serial number : 123��
```

Also note that the `0x00` in the middle of the hwSerialNum ends up NULL-terminating the whole thing, causing any following characters not to be printed.

### Steps to reproduce

Save the certificate above as a file and run the sample program: `programs/x509/cert_app mode=file filename=path/to/cert.crt`

### Additional information

I already implemented the hex conversion locally. If we agree on changing the format of the hardware serial number in `x509_info_subject_alt_name` to hex, I can submit a pull request.

### PR DESCRIPTION
**Pull request by Kabbah**
## Description
Change `x509_info_subject_alt_name` to render the serial numbers in HardwareModuleName SAN extensions (RFC 4108) in hex encoding.

Fixes #6262. More details about the reason of this change in the underlying issue.

## Status
**READY**

Backport 2.28: #6380

### CODE REVIEW

**Comment by davidhorstmann-arm** [comment]
Hi @Kabbah, thanks for your contribution!

Would you be able to rebase your first 2 commits on top of the `development` branch to avoid having merge commits in this PR?

**Comment by Kabbah** [comment]
> Hi @Kabbah, thanks for your contribution!
> 
> Would you be able to rebase your first 2 commits on top of the `development` branch to avoid having merge commits in this PR?

Done!

**Review by davidhorstmann-arm — DISMISSED** [review]
LGTM - thanks!

**Review by paul-elliott-arm — DISMISSED** [review]
LGTM

**Comment by gilles-peskine-arm** [comment]
We generally apply bug fixes to all supported versions of Mbed TLS: `development`, and the long-term support branches, currently `mbedtls-2.28`. It looks like the relevant code was already present in 2.28 so the bug fix should be backported.

In this case, it's possible that we might not want the bug fix, because there might be applications that have worked around the bug and that would be broken by the fix. I'm not sure how likely it is: I've only looked superficially at the bug description. @paul-elliott-arm @davidhorstmann-arm  If you think it's better to leave 2.28 alone, please argue why.

File: tests/data_files/Makefile [hunk 1]
```diff
@@ -276,6 +276,9 @@ all_final += server5-ss-forgeca.crt
 server5-othername.crt: server5.key
 	$(OPENSSL) req -x509 -new -subj "/C=UK/O=Mbed TLS/CN=Mbed TLS othername SAN" -set_serial 77 -config $(test_ca_config_file) -extensions othername_san -days 3650 -sha256 -key $< -out $@
```
Comment (gilles-peskine-arm) [inline tests/data_files/Makefile:278]: A bug fix in the library needs a non-regression test (present) and a [changelog entry](https://github.com/Mbed-TLS/mbedtls/blob/development/ChangeLog.d/00README.md) (missing). Please add a changelog entry.

**Review by gilles-peskine-arm — CHANGES_REQUESTED** [review]
Missing changelog

**Comment by davidhorstmann-arm** [comment]
@gilles-peskine-arm an omission on my part. I agree this needs a backport - it could conceivably break scripts that parse the output of `programs/cert_app` but it would be unlikely to break C programs using Mbed TLS (unless someone is doing something strange with the output of `mbedtls_x509_crt_info()`).

File: tests/data_files/Makefile (hunk 1 above)
Comment (Kabbah) [reply tests/data_files/Makefile:278]: Hello @gilles-peskine-arm, thank you for the reminder. I added a bugfix changelog entry as requested.

**Comment by Kabbah** [comment]
Thank you all for the reviews. At first I was unsure whether this would need a backport or not, because the bugfix requires a breaking change in the behavior of `mbedtls_x509_crt_info()` (even though it's simply an informational function). In the event we decide to backport this bugfix, should I reproduce the changes on top of the `mbedtls-2.28` branch and open a new pull request?

**Comment by davidhorstmann-arm** [comment]
@Kabbah that's correct, please open a new PR on top of `mbedtls-2.28` (usually we use the same name with "[Backport 2.28]" in front) then link to it from here.

**Review by davidhorstmann-arm — APPROVED** [review]
Looks good to me, thanks again! Please raise a backport on 2.28.

For future reference, when making new changes we'd prefer you to add new commits on the branch rather than force-pushing as it makes it easier to review new additions.

**Comment by Kabbah** [comment]
> Looks good to me, thanks again! Please raise a backport on 2.28.
> 
> For future reference, when making new changes we'd prefer you to add new commits on the branch rather than force-pushing as it makes it easier to review new additions.

Thank you. I just submitted the backport PR (#6380). The changes I force-pushed were very minor: a simple code style fix (one space) and a test description update, and I preferred to fixup my previous commits rather than making new ones to avoid polluting upstream.

**Comment by davidhorstmann-arm** [comment]
Commenting to update PR

**Comment by daverodgman** [comment]
@paul-elliott-arm looks like this has had a very light update since your approval - if you can give it a quick look & the backport also, it should be a very quick job to get this merged & stop it hanging around / lighten the CI load.

**Review by paul-elliott-arm — APPROVED** [review]
LGTM

**Review by paul-elliott-arm — APPROVED** [review]
Faithful backport

**Pull request merged**